### PR TITLE
feat: add a custom made treewide formatter

### PR DESCRIPTION
**Pull request by amadalusia**
#### Describe your PR, what does it fix/add?
~~This adds treefmt-nix to the Nix flake. treefmt brings together most formatters (including clang-format and alejandra) and brings its together into one. The main benefit that comes out of this is now you can just run one command `nix fmt .` and format the entire project. (Well, at least the .cpp and the .nix files, since they are the only file formats that matter here lol.)~~ This adds a new treewide formatter to the Hyprland flake, as designed by NotAShelf. This checks for any C++ or Nix file and formats it with clang-format (and maybe clang-tidy) for C++, and Alejandra, Deadnix and Statix for Nix files.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
~~Doesn't seem to be anything that seems like abhorrently terrible other than a new flake input making the download process like 0.05% slower. (THE HORROR...)~~

~~It shouldn't break anything Hyprland or the flake, the only change made was the formatter.~~ Thanks Statix and Deadnix for calling BS to this claim.

#### Is it ready for merging, or does it need work?
I believe it is ready.

### CODE REVIEW

**Comment by amadalusia** [comment]
cc @fufexan @NotAShelf

**Comment by NotAShelf** [comment]
I'm not sure if we really need treefmt-nix. *If* we had a convoluted stack of formatters, that could be useful but with just two I don't see a reason why we couldn't use a bash script to format files by filetype.

**Comment by amadalusia** [comment]
>_If_ we had a convoluted stack of formatters, that could be useful but with just two I don't see a reason why we couldn't use a bash script to format files by filetype.

I will agree in saying that this is kind of an overkill way of formatting with Nix in this case, but it is better for speed and it may help once we decide we should format Markdown files or other stuff too (probably not going to happen).

**Comment by NotAShelf** [comment]
here's a quick and relatively fast (much faster than bash) example:

```nix
writeDashBin "lint" {
  makeWrapperArgs = [
    "--prefix" "PATH" ":" "${lib.makeBinPath [ pkgs.alejandra ]}"
  ];
} ''
	if [ -z "''${1:-""}" ] || [ "$1" == "." ]; then
      fd '.*\.nix' . -x statix fix -- {} \;
      fd '.*\.nix' . -X deadnix -e -- {} \; -X nixfmt {} \;
    else
      statix fix -- "$1"
      deadnix -e "$1"
      nixfmt "$1"
    fi
''
```

apply clang-formatter as seen fit, and we have a treewide formatter without a module addition.

**Comment by amadalusia** [comment]
> apply clang-formatter as seen fit, and we have a treewide formatter without a module addition.

This script looks really interesting :eyes: I can think of some ways of expanding this.

**Comment by amadalusia** [comment]
```nix
{ writeDashBin, deadnix, statix, alejandra, llvmPackages_19, lib }:
writeDashBin "fmt" {
  makeWrapperArgs = [
    "--prefix" "PATH" ":" "${lib.makeBinPath [ deadnix statix alejandra llvmPackages_19.clang-unwrapped ]}"
  ];
} ''
  # thanks NotAShelf for the nix formatting script :)
  nix_format() {
    if [ -z "''${1:-""}" ] || [ "$1" == "." ]; then
      fd '.*\.nix' . -x statix fix -- {} \;
      fd '.*\.nix' . -X deadnix -e -- {} \; -X alejandra {} \;
    else
      statix fix -- "$1"
      deadnix -e "$1"
      alejandra "$1"
    fi
  }

  cpp_format() {
    if [ -z "''${1:-""}" ] || [ "$1" == "." ]; then
      fd '.*\.cpp' . -x clang-format -i {} \;
    else
      clang-format -i "$1"
    fi
  }

  for i in $@; do
    echo "formatting $i ..."
    nix_format
    cpp_format
  done
''
```

**Comment by NotAShelf** [comment]
I think you can use `clang-tools` for a smaller-closure package that provides clang-format.

**Comment by amadalusia** [comment]
I believe this is complete, the formatter should be good now.

File: nix/treefmt.nix
Comment (fufexan) [inline nix/treefmt.nix:1]: Leftover?

**Review by fufexan — COMMENTED** [review]
Looks good, and this is a nice addition. I think it would be complemented nicely by https://github.com/cachix/git-hooks.nix. @NotAShelf thoughts?

**Comment by NotAShelf** [comment]
> Looks good, and this is a nice addition. I think it would be complemented nicely by [cachix/git-hooks.nix](https://github.com/cachix/git-hooks.nix). @NotAShelf thoughts?

I'm in favor of *not* adding extra inputs, but I'd be okay git-hooks.nix since there's no good alternative.

File: nix/treefmt.nix
Comment (amadalusia) [reply nix/treefmt.nix:1]: yeah

**Comment by amadalusia** [comment]
ill see about integrating git-hooks, seems to be a popular move.

**Comment by amadalusia** [comment]
@vaxerski how would you feel about adding clang-tidy into the codebase?

**Comment by amadalusia** [comment]
> ill see about integrating git-hooks, seems to be a popular move.
basically done, commiting once its *actually* done

**Comment by amadalusia** [comment]
Barebones implemenetation of git-hooks made. Highly suggest a thorough look.

**Comment by amadalusia** [comment]
what did i mean by that

what i mean to say now is it makes the argument check simpler since if the script is called without args it will now know that.

**Comment by fufexan** [comment]
Two things I've noticed:
1. `nix fmt` is very slow, takes 24s on my machine.
2. `subprojects/` should not be formatted.

**Comment by amadalusia** [comment]
> Two things I've noticed:
> 
> 1. `nix fmt` is very slow, takes 24s on my machine.
> 2. `subprojects/` should not be formatted.

Is the former an initial run? It took like 3 or 5 seconds for me.
The latter was fixed in the commit above.

**Comment by fufexan** [comment]
Yeah, the first point was an initial run after I intentionally misformatted flake.nix.

**Comment by amadalusia** [comment]
I'll see if I can replicate any of the speed degredations that you get with the format tool.

**Comment by amadalusia** [comment]
image
image

**Comment by fufexan** [comment]
I figured out why it was slow as hell, and it was the subprojects. Currently works pretty well I'd say.

**Comment by amadalusia** [comment]
fair enough lol, thanks for the fix

**Review by amadalusia — COMMENTED** [review]
Im happy with that yeah (github wont stop bothering)

File: nix/formatter.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
+{
+  writers,
+  deadnix,
+  statix,
+  alejandra,
+  llvmPackages_19,
+  fd,
+  lib,
+}:
+writers.writeDashBin "hyprland-treewide-formatter" {
```
[the remark this answers is not in the document]
Comment (amadalusia) [reply nix/formatter.nix:10]: took around the same time as writeDashBin, safe to say it works! i think i can just shebang it to bash anyway.

File: flake.nix [hunk 2]
```diff
@@ -55,6 +55,8 @@
       inputs.hyprutils.follows = "hyprutils";
       inputs.hyprwayland-scanner.follows = "hyprwayland-scanner";
     };
+
+    pre-commit-hooks.url = "github:cachix/git-hooks.nix";
```
[the remark this answers is not in the document]
Comment (amadalusia) [reply flake.nix:59]: done!

File: nix/formatter.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (NotAShelf) [reply nix/formatter.nix:10]: lets go with writeShellApplication then (no dash)

[the remark this answers is not in the document]
Comment (amadalusia) [reply nix/formatter.nix]: didn't see this until now, but yeah removed dash

**Comment by vaxerski** [comment]
> how would you feel about adding clang-tidy into the codebase? i feel may its best be left in another PR but your mood is the deciding factor here

no.

**Comment by amadalusia** [comment]
> no.

thats fair enough

**Comment by fufexan** [comment]
Perhaps another nice addition would be a formatting check in CI. Not sure how to achieve this without adding a --check flag to the formatter script.

But that can be done in another PR. For now I'll merge this.

**Review by fufexan — APPROVED** [review]
LGTM, thanks for the great work!

**Comment by NotAShelf** [comment]
I'll take a look at implementing a check script.

**Pull request merged**